### ISSUE TITLE
memory leak, possibly related to Grizzly HTTP/2 implementation

### ISSUE THREAD

**Issue by maxdreamland** [payara/Payara#2872]
# TL;DR #
----------

We found, what we believe is a memory leak, possibly related to HTTP/2 implementation

# Description #
----------------
We recently switched from Payara 4 to Payara 5 and suddenly had problems with running out of heap memory. We first analyzed the garbage collection behaviour over several hours and found that the old gen heap was increasingly getting larger. With full gc runs becoming more and more frequent.

A diagram of the GC behaviour can be seen here (the red marks are the timestamps where full gc runs occured):

https://www.dropbox.com/s/8vhye91cq2xbmsr/heap_after_gc.jpg?dl=0

After that we analyzed heap dumps to find out where the heap was spend, and we found that more than half our heap size (~266 MB of 512 MB of heap) was occupied by byte[] instances belonging to sun.security.ssl.EngineInputRecord.

The full stack traces can be seen in this image:

https://www.dropbox.com/s/9qua02igpuc1jj7/stacktrace.JPG?dl=0

The stacktrace seemed to imply that the Grizzly HTTTP/2 - ALPN implementation might be the culprit. Then we noticed that in our local development Payara 5 installation we had the following setting as part of the domain.xml for all the http listeners:

`<http http2-enabled="false" server-header="false" xpowered-by="false" max-connections="250" default-virtual-server="server">`

And that entry was missing in our AWS production Payara 5 instance. We assumed that the default would be to activate the feature and therefore added the missing entry to our production domain.xml as well, which fixed the memory leak problem.

To confirm that the memory leak was indeed related to the underlying HTTP implementation and not our business code, we made a stress test with a simple REST service (via HTTPS) that just returns a static string and does nothing else. And we were able to see the same accumulation of byte[] (from the same sources mentioned above) on the heap. This should also be a simple way to reproduce the problem.

## Expected Outcome

more or less stable use of heap memory

## Current Outcome

old gen heap memory will continously accumulate byte[] instances belonging to the Grizzly HTTP/2 implementation

## Steps to reproduce (Only for bug reports) 

implement a simple webservice, for example via Jersey:

```
	@GET
	@Consumes(MediaType.TEXT_PLAIN + ";charset=utf-8")
	@Path("environment")
	public String getEnvironment() {
		return "development";
	}

```

and then repeatedly call it via HTTPS. you probably need to do this several thousand times to see significant accumulation of heap space.

## Environment ##

- **Payara Version**: 5.181
- **Edition**: Full
- **JDK Version**: 1.8.0_162
- **Operating System**: Linux 4.14.33-51.37.amzn1.x86_64

**Comment by MattGill98** [comment]
Hi @maxdreamland,

This looks very similar to the issue addressed in https://github.com/payara/Payara/pull/2831. You can always test this by building the master branch and testing on that. Alternatively, support customers get access to patch releases to fix bugs such as these. If it is the same issue, the fix is scheduled for the 183 release.

Hope this helps,

Matt

**Comment by maxdreamland** [comment]
Thank you for the update, Matt.

**Comment by smillidge** [comment]
Fixed by #2831
